lib/rex/proto/ldap/server.rb

```diff
@@ -228,25 +285,50 @@ def encode_ldap_response(msgid, code, dn, msg, tag)
         # @param attrflt [Array, Symbol] LDAP attribute filter
         #
         # @return [Array] Query matches
-        def search_ldif(filter, msgid, attrflt = :all)
-          return [] if @ldif.nil? || @ldif.empty?
-
-          ldif.map do |dn, entry|
-            next unless filter.match(entry)
 
+        def search_result(filter, msgid, attrflt = :all)
+          if @ldif.nil? || @ldif.empty?
             attrs = []
-            entry.each do |k, v|
-              if attrflt == :all || attrflt.include?(k.downcase)
-                attrvals = v.map(&:to_ber).to_ber_set
-                attrs << [k.to_ber, attrvals].to_ber_sequence
+            if attrflt.is_a?(Array)
+              attrflt.each do |at|
+                attrval = [Rex::Text.rand_text_alphanumeric(10)].map(&:to_ber).to_ber_set
+                attrs << [at.to_ber, attrval].to_ber_sequence
               end
+              dn = "dc=#{Rex::Text.rand_text_alphanumeric(10)},dc=#{Rex::Text.rand_text_alpha(4)}"
+              appseq = [
+                dn.to_ber,
+                attrs.to_ber_sequence
+              ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
+              [msgid.to_ber, appseq].to_ber_sequence
             end
-            appseq = [
-              dn.to_ber,
-              attrs.to_ber_sequence
-            ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
-            [msgid.to_ber, appseq].to_ber_sequence
-          end.compact
+          else
+            ldif.map do |bind_dn, entry|
+              next unless filter.match(entry)
+
+              attrs = []
+              entry.each do |k, v|
+                if attrflt == :all || attrflt.include?(k.downcase)
+                  attrvals = v.map(&:to_ber).to_ber_set
+                  attrs << [k.to_ber, attrvals].to_ber_sequence
+                end
+              end
+              appseq = [
+                bind_dn.to_ber,
+                attrs.to_ber_sequence
+              ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
+              [msgid.to_ber, appseq].to_ber_sequence
+            end.compact.join
+          end
+        end
+
+        #
+        # Sets the tasks to be performed after processing of pdu object
+        #
+        # @param proc [Proc] block of code to execute
+        #
+        # @return pdu_process [Proc] steps to be executed
+        def processed_pdu_handler(&proc)
+          @pdu_process = proc if block_given?
```

Comment:
Not sure if this is the right place to warp the call in a `proc`.  The signature expectations for a handler end up less strictly defined here.

Consider:
```suggestion
        def processed_pdu_handler(processed_data)
          @pdu_process.call(processed_data) if @pdu_process
```

What layer to wrap the call in a `proc` at is somewhat subjective so happy to discuss and iterate on this approach. My concern is to determine if there is some reason it is better expect this to be a `proc` vs wrapping it again and expecting a block to have been passed.